revscoring/dependent.py

```diff
@@ -85,15 +85,32 @@ def _solve(dependent, cache, history=None):
                 # No dependencies?  OK.  Let's try that.
                 dependencies = []
 
-            # Generate args from dependencies
-            values = []
+            # Generate args for process function from dependencies (if any)
+            args = []
             for dependency in dependencies:
                 value, cache, history = _solve(dependency, cache, history)
-                values.append(value)
+                args.append(value)
 
             # Generate value
-            value = dependent(*values)
+            value = dependent(*args)
 
             # Add value to cache
             cache[dependent] = value
             return cache[dependent], cache, history
+
+
+def draw(dependent, cache=None, depth=0):
+    print("\t" * depth + " - " + str(dependent))
+
+    cache = cache or {}
+
+    # Check if we're a dependent with explicit dependencies
+    if hasattr(dependent, "dependencies"):
+        for dependency in dependent.dependencies:
+            if dependency not in cache:
+                draw(dependency, depth=depth+1)
+            else:
+                draw("CACHED", depth=depth+1)
```

Comment:
Indeed.  See the example below.  

    $ python
    Python 3.4.0 (default, Apr 11 2014, 13:05:11) 
    [GCC 4.8.2] on linux
    Type "help", "copyright", "credits" or "license" for more information.
    >>> from revscoring.dependent import draw
    >>> from revscoring.features import diff
    >>> from revscoring.datasources import revision, parent_revision
    >>> draw(diff.added_badwords_ratio, cache={revision.text: "test", parent_revision.text: "test"})
     - <((diff.badwords_added / max(diff.words_added, 1)) / max((parent_revision.badwords / max(parent_revision.words, 1)), 1))>
    	 - <(diff.badwords_added / max(diff.words_added, 1))>
    		 - <diff.badwords_added>
    			 - <is_badword>
    			 - <diff.added_words>
    				 - <diff.added_segments>
    					 - <diff.operations>
    						 - CACHED
    						 - CACHED
    		 - <max(diff.words_added, 1)>
    			 - <diff.words_added>
    				 - <diff.added_words>
    					 - <diff.added_segments>
    						 - <diff.operations>
    							 - CACHED
    							 - CACHED
    			 - <1>
    	 - <max((parent_revision.badwords / max(parent_revision.words, 1)), 1)>
    		 - <(parent_revision.badwords / max(parent_revision.words, 1))>
    			 - <parent_revision.badwords>
    				 - <is_badword>
    				 - <parent_revision.words>
    					 - CACHED
    			 - <max(parent_revision.words, 1)>
    				 - <parent_revision.words>
    					 - <parent_revision.words>
    						 - CACHED
    				 - <1>
    		 - <1>